File: pset6/sentimental-credit/credit.py
from cs50 import get_string


def luhns(creditNumber):
    sum = 0
    flag = False
    for index in range(len(creditNumber) - 1, -1, -1):
        if flag:
            number = int(creditNumber[index]) * 2
            if(len(str(number)) == 1):
                sum += number
            else:
                stringedNum = str(number)
                for i in range(0, len(stringedNum)):
                    sum += int(stringedNum[i])
            flag = False
        else:
            sum += int(creditNumber[index])
            flag = True

    if(str(sum)[-1] == '0'):
        return True
    else:
        return False


def main():

    creditNumber = get_string("Number: ")

    firstDigit, secondDigit = creditNumber[0], creditNumber[1]
    length = len(creditNumber)
    if luhns(creditNumber):
        if length == 13 and firstDigit == "4":
            print("VISA\n")
        elif length == 15 and (firstDigit + secondDigit) == "34" or (firstDigit + secondDigit) == "37":
            print("AMEX\n")
        elif length == 16:
            if firstDigit == "4":
                print("VISA\n")
            else:
                for i in range(51, 56):
                    if(firstDigit + secondDigit == str(i)):
                        print("MASTERCARD\n")
                        break
                    print("INVALID\n")
    else:
        print("INVALID\n")


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>


int main(void)
{
    long creditNumber = get_long("Number: ");

    int length = 0;
    long checkLength = creditNumber;
    while (checkLength != 0)
    {
        checkLength /= 10;
        length++;
    }

    int counter = length - 2;
    long firstTwoNum = creditNumber;
    while (counter != 0)
    {
        firstTwoNum /= 10;
        counter--;
    }

    long lastNum = creditNumber;
    int sumOfTimesTwo = 0;
    int sumOfOther = 0;
    bool isNeeded = false;
    while (creditNumber > 0)
    {
        creditNumber = lastNum / 10;
        if (isNeeded)
        {
            int currNum = (lastNum % 10) * 2;
            if (currNum >= 10)
            {
                sumOfTimesTwo += currNum / 10;
                sumOfTimesTwo += currNum % 10;
            }
            else
            {
                sumOfTimesTwo += currNum;
            }
            isNeeded = false;
        }
        else
        {
            sumOfOther += lastNum % 10;
            isNeeded = true;
        }

        lastNum = creditNumber;
    }

    int sum = sumOfTimesTwo + sumOfOther;

    if (sum % 10 == 0)
    {
        if (length == 15 && (firstTwoNum == 34 || firstTwoNum == 37))
        {
            printf("AMEX\n");
        }
        else if ((length == 13 || length == 16) && (firstTwoNum / 10) == 4)
        {
            printf("VISA\n");
        }
        else if (length == 16 && (firstTwoNum == 51 || firstTwoNum == 52 || firstTwoNum == 53 || firstTwoNum == 54 || firstTwoNum == 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}